### DISCUSSION TITLE
ST7789 128x128 on M5stackS3 partially displaying

### DISCUSSION THREAD

**Discussion by ouinouin** [arendst/Tasmota#20413] [Q&A]
Hi, i began to play with the M5stack atomS3 , with the 0,85" display 128x128 , 
as i could get from esphome forum , this is driven by the ST7789v driver , so i use the universal driver on a precompiled tasmota lvgl.
I managed to setup all the pins, option A3 , and gave a try with this display.ini : 
```:H,ST7789,128,128,16,SPI,1,*,*,*,*,*,*,*,40
:S,2,1,3,0,30,30
:I
01,A0
11,A0
3A,81,55
36,81,00
21,80
13,80
29,A0
:o,28
:O,29
:A,2A,2B,2C
:R,36
:0,C0,00,50,00
:1,A0,50,00,01
:2,00,00,00,02
:3,60,00,00,03
:i,20,21
#
```

unfortunately, the image looks shifted, at boot the tasmota logois virtually located under the bottom of the screen, 
i tried all the possible rotations and to my surprise i got different results.
is there some offset or something as i could see on esphome code, what kind of magic should i invoke to get the rotation 0 displayed correctly ? 
maybe @gemu2015 could give me a hint :-) 
displayrotate 0 
![20240106_164134](https://github.com/arendst/Tasmota/assets/6713759/2fb1ec61-f9c5-42ff-956f-c54819e36bcc)

displayrotate 1 
![20240106_164031](https://github.com/arendst/Tasmota/assets/6713759/2dae6029-abac-4017-a7c1-b6390ac1b025)
displayrotate 2
![20240106_163940](https://github.com/arendst/Tasmota/assets/6713759/860d8cef-a0e1-4f39-81f7-947448f40bf7)
displayrotate 3
![20240106_164048](https://github.com/arendst/Tasmota/assets/6713759/b5763961-f914-4837-9211-4185f0b99a7f)

edit : while digging : https://github.com/m5stack/M5GFX/blob/master/src/M5GFX.cpp :+1: 
```
if (board == 0 || board == board_t::board_M5AtomS3)
      {
        _pin_reset(GPIO_NUM_34, use_reset); // LCD RST
        bus_cfg.pin_mosi = GPIO_NUM_21;
        bus_cfg.pin_miso = GPIO_NUM_13;
        bus_cfg.pin_sclk = GPIO_NUM_17;
        bus_cfg.pin_dc   = GPIO_NUM_33;
        bus_cfg.spi_mode = 0;
        bus_cfg.spi_3wire = true;
        bus_spi->config(bus_cfg);
        bus_spi->init();
        id = _read_panel_id(bus_spi, GPIO_NUM_15);
        if ((id & 0xFFFFFF) == 0x079100)
        {  //  check panel (GC9107)
          board = board_t::board_M5AtomS3;
          ESP_LOGW(LIBRARY_NAME, "[Autodetect] board_M5AtomS3");
          bus_spi->release();
          bus_cfg.spi_host = SPI3_HOST;
          bus_cfg.freq_write = 40000000;
          bus_cfg.freq_read  = 16000000;
          bus_spi->config(bus_cfg);
          bus_spi->init();
          auto p = new Panel_GC9107();
          p->bus(bus_spi);
          {
            auto cfg = p->config();
            cfg.pin_cs  = GPIO_NUM_15;
            cfg.pin_rst = GPIO_NUM_34;
            cfg.panel_width = 128;
            cfg.panel_height = 128;
            cfg.offset_y = 32;
            cfg.readable = false;
            p->config(cfg);
          }
          _panel_last.reset(p);
          _set_pwm_backlight(GPIO_NUM_16, 7, 256, false, 48);

          goto init_clear;
        }
        lgfx::pinMode(GPIO_NUM_48, lgfx::pin_mode_t::input); // LCD RST
        bus_spi->release();
      }
```
they deal with a panel , GC1907... , with a 32 pixel offset on the Y , most likely ther reforged something similar to  the ST 7789..., just with a narrower size not really supported by the ST7789 driver , so they clip the image with an offset or something like that. comments welcome

**Comment by ouinouin** [comment]
https://github.com/Bodmer/TFT_eSPI/issues/2072
apparently this could be the same display as on the liligo T-QT
i gave a try to the 7735 ini files, i had more luck, the display is nearly aligned on rotation 0

```
:H,ST7735,128,128,16,SPI,1,*,*,*,*,*,*,*,40
:S,2,1,1,0,40,20
:I
EF,3,03,80,02
CF,3,00,C1,30
ED,4,64,03,12,81
E8,3,85,00,78
CB,5,39,2C,00,34,02
F7,1,20
EA,2,00,00
C0,1,23
C1,1,10
C5,2,3e,28
C7,1,86
36,1,48
37,1,00
3A,1,55
B1,2,00,18
B4,1,04
B6,3,08,82,27
F2,1,00
26,1,01
E0,0F,0F,31,2B,0C,0E,08,4E,F1,37,07,10,03,0E,09,00
E1,0F,00,0E,14,03,11,07,31,C1,48,08,0F,0C,31,36,0F
11,80
29,80
:o,28
:O,29
:A,2A,2B,2C,16
:R,36
:0,C0,00,00,00
:1,A0,00,00,01
:2,00,00,00,02
:3,60,00,00,03
:i,20,21
#
#
```
here the result with this display.ini file and the lvgl example quickly and dirtily copied. mind that with displaydimmer 50 , the power is not enough to illuminate de screen.
displaydimmer 60 give a confortable view.

![20240106_182444](https://github.com/arendst/Tasmota/assets/6713759/5e3fd064-44b8-4200-9b96-a21426345ab3)

i ll try to make an autoconfig for the Atom S3 with all those infos gathered.

**Reply by ouinouin** [reply]
i checked deeply into many examples, to finally get a good result with this : 
```
:H,ST7735S,128,128,16,SPI,1,*,*,*,*,*,*,*,40

:I
01,80
11,E0
B1,3,05,3C,3C
B2,3,05,3C,3C
B3,6,05,3C,3C,05,3C,3C
B4,1,07
C0,3,A4,04,84
C1,1,C5
C2,2,0D,00
C3,2,8D,6A
C4,2,8D,EE
C5,1,21
FC,1,80
F0,1,11
D6,1,CB

20,0
36,1,C8
3A,1,05



2A,4,00,02,00,81
2B,4,00,01,00,80

E0,10,0D,0C,0C,0E,0E,00,00,00,00,09,23,31,00,0C,03,1A
E1,10,0A,05,06,07,08,01,00,00,00,05,20,2E,00,0A,01,1A



13,A0
29,E0

:o,28
:O,29
:A,2A,2B,2C,16
:R,36
;CC_A8_08__68
:0,CC,02,01,00
:1,68,01,02,01
:2,08,02,01,02
:3,A8,01,02,03
:i,21,20
:i,21,20
:B10,0
#
```

colors are no more inverted , the file is basically the one from the M5stick , but with the DMA disabled  :B10,0 instead of :B10,1 , i adjusted and checked the offsets for the rotations, picture looks ok on all rotation modes.

**Reply by ouinouin** [reply]
diggin still a bit more, i can see that there i a lot of unddocumenterd registers for the GC9107.

i inspired from  from the M5stack graphics code https://github.com/m5stack/M5GFX/blob/master/src/lgfx/v1/panel/Panel_GC9A01.hpp line 94 an then got this display.ini working : 

`:H,ST7735S,128,128,16,SPI,1,*,*,*,*,*,*,*,40

:I
01,80
11,E0
EF,80,5
B0,1,C0
B2,1,2F
B3,1,03
B6,1,19
B7,1,01
AC,1,CB
AB,1,0E
B4,1,04
A8,1,19
B8,1,08
E8,1,24
E9,1,48
EA,1,22
C6,1,30
C7,1,18
3A,1,05
F0,14,1F,28,04,3E,2A,2E,20,00,0C,06,00,1C,1F,0f
F1,14,00,2D,2F,3C,6F,1C,0B,00,00,00,07,0D,11,0f
29,0
:o,28
:O,29
:A,2A,2B,2C,16
:R,36
:0,CC,02,01,00
:1,68,01,02,01
:2,08,02,01,02
:3,A8,01,02,03
:i,21,20
:i,21,20
;B10,0
#`

interresting point is that the gamma registers are matching the documentation (i could honestly see no effects on my tests.
there seems to be a lot of useless and undocumented registers, if someone has extensive documentation on this display...

**Reply by s-hadinger** [reply]
The easiest approach is to start from existing code. It's really hard to do it from the data-sheet, and you may not know all the details of the device.

Is the above working? If yes, can you do a pull request?

**Reply by ouinouin** [reply]
@s-hadinger , thanks for comment.
yes both of those are working, i ll give a try for a pull request within this week.

**Comment by goodkiller** [comment]
@ouinouin please share your template? I'm not able to get it work.

**Comment by ouinouin** [comment]
@goodkiller my template was published in the screens list on tasmota sourcecode.

**Reply by sfromis** [reply]
I suppose that you refer to these:
https://github.com/arendst/Tasmota/blob/development/tasmota/displaydesc/ST7735S_AtomS3_display.ini
https://github.com/arendst/Tasmota/blob/development/tasmota/displaydesc/ST7735S_AtomS3_SPI2_display.ini
(Technically, they are not part of the source code Tasmota is built from, that directory is more of a collection of sample display descriptors).

**Reply by goodkiller** [reply]
@ouinouin I asked tasmota template (command "template") not ini configuration.

**Reply by ouinouin** [reply]
@goodkiller : here the command result : 
`17:04:35.780 RSL: RESULT = {"NAME":"ATOMS3DISPLAY","GPIO":[0,1,1,0,1056,1,1,1,6210,0,0,0,0,0,0,768,992,736,0,0,0,704,800,1024,0,0,0,0,416,0,32,0,0,0,0,0,0,0],"FLAG":0,"BASE":1}
`
![atoms3config template 2](https://github.com/user-attachments/assets/55c9136b-196e-4778-ab0d-e4f5fbed68c8)
![template setup M5atomS3](https://github.com/user-attachments/assets/ba86b5e6-14b9-40bc-9009-f10c95564ad2)
hope this helps

**Reply by Patrick197030100** [reply]
hello, since the introduction of LVGL esphom8.0 I no longer display LCD on my atom s3 NO GOOD FOR LCD

**Reply by sfromis** [reply]
As this Github project is about using it with Tasmota, you'll have to post about your issue where it is on-topic.
